lib/src/command/add.dart

```diff
@@ -217,46 +315,48 @@ class AddCommand extends PubCommand {
 
       packageRange = cache.sources['git']
           .parseRef(packageName, git)
-          .withConstraint(VersionConstraint.any);
+          .withConstraint(constraint ?? VersionConstraint.any);
       pubspecInformation = {'git': git};
     } else if (path != null) {
       packageRange = cache.sources['path']
           .parseRef(packageName, path, containingPath: entrypoint.pubspecPath)
-          .withConstraint(VersionConstraint.any);
+          .withConstraint(constraint ?? VersionConstraint.any);
       pubspecInformation = {'path': path};
     } else if (sdk != null) {
       packageRange = cache.sources['sdk']
           .parseRef(packageName, sdk)
           .withConstraint(constraint ?? VersionConstraint.any);
       pubspecInformation = {'sdk': sdk};
-      if (constraint != null) {
-        pubspecInformation['version'] = constraint.toString();
-      }
     } else {
       final hostInfo =
           hasHostOptions ? {'url': hostUrl, 'name': packageName} : null;
 
       if (hostInfo == null) {
         pubspecInformation = constraint?.toString();
-      } else if (constraint == null) {
-        pubspecInformation = {'hosted': hostInfo};
       } else {
-        pubspecInformation = {
-          'hosted': hostInfo,
-          'version': constraint.toString()
-        };
+        pubspecInformation = {'hosted': hostInfo};
       }
 
       packageRange = PackageRange(packageName, cache.sources['hosted'],
           constraint ?? VersionConstraint.any, hostInfo ?? packageName);
     }
 
+    if (pubspecInformation is Map && constraint != null) {
```

Comment:
Not sure if this is necessary?
```suggestion
    if (pubspecInformation is Map && constraint != null && constraint != VersionConstraint.any) {
```

Reply:
hm I'm open to adding the condition, but because of the way we parsed it above:
```dart
constraint = splitPackage.length == 2
    ? VersionConstraint.parse(splitPackage[1])
     : null;
```
`constraint` will only take on `VersionConstraint.any` if it is explicitly passed in, so in that case it becomes a question of whether we want to preserve it if explicitly passed in. The one main place I think where it makes a difference is when adding packages hosted on a different pub server (e.g. `pub add foo:any --hosted-url https://pub.foo`), and whether we want it to reflect:

```
foo: 
  version: any
  hosted: 
    name: foo
    url: https://pub.foo
```

or 

```
foo: 
  version: ^1.2.3
  hosted: 
    name: foo
    url: https://pub.foo
```

It currently returns the former!